index.js

```diff
@@ -3598,23 +3598,24 @@
   //. > S.intercalate (String) (' mississippi, ') ([])
   //. ''
   //. ```
-  function intercalate(monoid) {
+  function intercalate(Monoid) {
     return function(sep) {
-      return function(arr) {
-        if (arr.length === 0) return Z.empty (monoid);
-
-        var head = arr[0];
-        var tail = arr.slice (1);
-        return Z.concat (
-          head,
-          Z.foldMap (monoid, function(x) { return Z.concat (sep, x); }, tail)
+      return function(foldable) {
+        var empty = true;
+        return Z.reduce (
+          function(monoid, x) {
+            return Z.concat (monoid,
+                             empty ? (empty = false, x) : Z.concat (sep, x));
```

Comment:
Would this sort of side effect work well with e.g. trees (which fold recursively)?

Reply:
I'm not sure. Let's add a test case. :)

If necessary we can change the type of the accumulator from `a` to `{ empty :: Boolean, value :: a }`. I avoided this due to performance concerns, but correctness is far more important than performance.

Reply:
Nice one!

Reply:
This side effect makes me nervous. We have been bitten in the past by similar code.

I would prefer the following:

```javascript
var result = Z.reduce (
  function(acc, x) {
    return {
      empty: false,
      value: Z.concat (acc.value, acc.empty ? x : Z.concat (sep, x))
    };
  },
  {empty: true, value: Z.empty (Monoid)},
  foldable
);
return result.value;
```